core/types/header_ext_test.go

```diff
@@ -0,0 +1,244 @@
+
+	extra = &HeaderExtra{
+		ExtDataHash: [32]byte{1},
+	}
+	SetHeaderExtra(h, extra)
+
+	extra = GetHeaderExtra(h)
+	assert.Equal(t, &HeaderExtra{
+		ExtDataHash: [32]byte{1},
+	}, extra)
+}
+
+func TestHeaderExtraRLP(t *testing.T) {
+	t.Parallel()
+
+	eth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+	}
+	extra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+
+	writer := bytes.NewBuffer(nil)
+	err := extra.EncodeRLP(eth, writer)
+	require.NoError(t, err)
+
+	stream := rlp.NewStream(bytes.NewReader(writer.Bytes()), 0)
+	decodedExtra := new(HeaderExtra)
+	decodedEth := new(ethtypes.Header)
+	err = decodedExtra.DecodeRLP(decodedEth, stream)
+	require.NoError(t, err)
+
+	wantEth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		Difficulty: new(big.Int),
+		Number:     new(big.Int),
+		Extra:      []byte{},
+	}
+	assert.Equal(t, wantEth, decodedEth)
+
+	wantExtra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+	assert.Equal(t, wantExtra, decodedExtra)
+}
+
+func TestHeaderExtraJSON(t *testing.T) {
+	t.Parallel()
+
+	eth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		// Required json fields
+		Difficulty: big.NewInt(2),
+		Number:     big.NewInt(3),
+	}
+	extra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+
+	encoded, err := extra.EncodeJSON(eth)
+	require.NoError(t, err)
+
+	decodedExtra := new(HeaderExtra)
+	decodedEth := new(ethtypes.Header)
+	err = decodedExtra.DecodeJSON(decodedEth, encoded)
+	require.NoError(t, err)
+
+	wantEth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		Difficulty: big.NewInt(2),
+		Number:     big.NewInt(3),
+		Extra:      []byte{},
+	}
+	assert.Equal(t, wantEth, decodedEth)
+
+	wantExtra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+	assert.Equal(t, wantExtra, decodedExtra)
+}
+
+func TestHeaderSerializable_updates(t *testing.T) {
+	t.Parallel()
+
+	t.Run("from", func(t *testing.T) {
+		t.Parallel()
+
+		eth := &ethtypes.Header{
+			ParentHash:       common.Hash{1},
+			UncleHash:        common.Hash{2},
+			Coinbase:         common.Address{3},
+			Root:             common.Hash{4},
+			TxHash:           common.Hash{5},
+			ReceiptHash:      common.Hash{6},
+			Bloom:            Bloom{7},
+			Difficulty:       big.NewInt(8),
+			Number:           big.NewInt(9),
+			GasLimit:         10,
+			GasUsed:          11,
+			Time:             12,
+			Extra:            []byte{13},
+			MixDigest:        common.Hash{14},
+			Nonce:            BlockNonce{15},
+			BaseFee:          big.NewInt(16),
+			WithdrawalsHash:  &common.Hash{17},
+			BlobGasUsed:      ptrTo(uint64(18)),
+			ExcessBlobGas:    ptrTo(uint64(19)),
+			ParentBeaconRoot: &common.Hash{20},
+		}
+		extras := &HeaderExtra{
+			ExtDataHash:    common.Hash{21},
+			ExtDataGasUsed: big.NewInt(22),
+			BlockGasCost:   big.NewInt(23),
+		}
+
+		got := new(HeaderSerializable)
+		got.updateFromEth(eth)
+		got.updateFromExtras(extras)
+
+		allFieldsAreSet(t, got)
+
+		want := &HeaderSerializable{
```

Comment:
This approach is error-prone in that it requires a reader to very closely inspect every single field and value (it's nice that the values increment, but the field order is still problematic). If you test via RLP (and/or JSON) round-tripping then you can simply assert that you get the exact input-struct values back. It will also cover both "to" and "from" sub tests.

Reply:
Good point, done in last commit 👍